Gsi update

mark-a-potts opened the pull request:
This PR references [https://github.com/NOAA-EMC/NCEPLIBS/issues/81](url)

I have made updates that I think are consistent with the other NCEPLIBS in the umbrella build as a starting point. I also introduced an approach to compiler-independent flags with variables set for functionality (e.g. using 8 byte reals) and flags set for each supported compiler (GNU and Intel, currently) based on the way that NASA does things for GEOS. I am open to not using that at this point, but I think if we can get a common set of flags across NCEP, it will simplify building with various compilers going forward.

kgerheiser commented on CMakeLists.txt:
```diff
@@ -99,30 +105,81 @@ foreach(kind ${kinds})
+
+  add_library(${lib_name} STATIC $<TARGET_OBJECTS:${lib_name}_f> $<TARGET_OBJECTS:${lib_name}_c>)
+
+  set_target_properties(${lib_name} PROPERTIES OUTPUT_NAME "${versioned_lib_name}")
```
We've been getting rid of library names with versions in them

kgerheiser commented on CMakeLists.txt:
```diff
@@ -99,30 +105,81 @@ foreach(kind ${kinds})
+  NAMESPACE ${PROJECT_NAME}::
+  FILE ${PROJECT_NAME}-targets.cmake)
+
+# bacio-config.cmake
```
`bacio`, I would just git rid of these comments

kgerheiser commented:
I kinda like the CMake flags thing.

Do you think we should put them into the CMakeModules repo and maintain them centrally?

kgerheiser commented:
Do you know what the purpose of all these variations of the library with `_DA`? It really makes the build system confusing.

Otherwise it built for me.

kgerheiser commented on CMakeLists.txt:
```diff
@@ -99,30 +105,81 @@ foreach(kind ${kinds})
-  
-  set_source_files_properties(${fortran_src} PROPERTIES
+  add_library(${lib_name}_f OBJECT ${fortran_src})
+  set_target_properties(${lib_name}_f PROPERTIES POSITION_INDEPENDENT_CODE ON)
```
Do you need this?

aerorahul replied:
No

kgerheiser commented on CMakeLists.txt:
```diff
@@ -25,49 +51,31 @@ STRING(COMPARE EQUAL ${CMAKE_BUILD_TYPE} "DEBUG" BUILD_DEBUG)
+if(IntelComp)
+      set(c_DA_allocation_def "DYNAMIC_ALLOCATION")
+      set(c_nonDA_allocation_def "STATIC_ALLOCATION")
+      message("Intel Compiler flags set")
```
I would remove these messages

kgerheiser commented on CMakeLists.txt:
```diff
@@ -25,49 +51,31 @@ STRING(COMPARE EQUAL ${CMAKE_BUILD_TYPE} "DEBUG" BUILD_DEBUG)
+set(c_8_definitions "F77_INTSIZE_8")
+set(c_d_flags "-DUNDERSCORE")
+set(fortran_4_flags)
+if(IntelComp)
```
Can this `if(IntelComp)` and the other compiler `comp`s general be removed and move this into the `if` at the top where you detect the compiler?

kgerheiser replied:
I guess you do need to know the compiler later on.

kgerheiser commented on CMakeLists.txt:
```diff
@@ -1,22 +1,48 @@
+    VERSION ${pVersion}
+    LANGUAGES C Fortran)
+
 set(${PROJECT_NAME}_VERSION ${PROJECT_VERSION} CACHE INTERNAL "${PROJECT_NAME} version number")
```
I think you can get rid of this.

kgerheiser commented:
Can you get rid of the `DoubleDA`, etc folders? Those aren't used anymore.

aerorahul commented on CMakeLists.txt:
```diff
@@ -25,49 +51,31 @@ STRING(COMPARE EQUAL ${CMAKE_BUILD_TYPE} "DEBUG" BUILD_DEBUG)
+set(c_DA_allocation_def "DYNAMIC_ALLOCATION")
+set(c_nonDA_allocation_def "STATIC_ALLOCATION")
+
+set(c_4_DA_flags "-DUNDERSCORE")
```
This is handled generically. Remove

aerorahul commented on CMakeLists.txt:
```diff
@@ -1,22 +1,48 @@
-if(CMAKE_CXX_COMPILER_ID STREQUAL "Intel")
+if(CMAKE_C_COMPILER_ID MATCHES "^(Intel)$")
+  include(${CMAKE_CURRENT_SOURCE_DIR}/cmake/Intel.cmake)
   set(IntelComp true )
```
Please remove intelcomp , gnucomp, etc

aerorahul commented on cmake/GNU.cmake:
```diff
@@ -0,0 +1,68 @@
+if (CMAKE_Fortran_COMPILER_VERSION VERSION_LESS 5.3)
```
This file is largely unnecessary. Same goes for intel.

DusanJovic-NOAA commented on CMakeLists.txt:
```diff
@@ -1,22 +1,48 @@
+if(APPLE)
+  add_compile_definitions(APPLE)
+elseif(UNIX)
+  add_compile_definitions(LINUX)
```
I do not see where ``APPLE`` and ``LINUX`` definitions are used. If they are not remove them.

DusanJovic-NOAA commented on CMakeLists.txt:
```diff
@@ -1,22 +1,48 @@
+
 STRING(COMPARE EQUAL ${CMAKE_BUILD_TYPE} "RelWithDebInfo" BUILD_RELEASE)
 STRING(COMPARE EQUAL ${CMAKE_BUILD_TYPE} "RELEASE" BUILD_RELEASE)
 STRING(COMPARE EQUAL ${CMAKE_BUILD_TYPE} "PRODUCTION" BUILD_PRODUCTION)
```
I do not see where ``BUILD_RELEASE`` and ``BUILD_PRODUCTION`` variables are used. If they are not remove them.

DusanJovic-NOAA commented on CMakeLists.txt:
```diff
@@ -25,49 +51,31 @@ STRING(COMPARE EQUAL ${CMAKE_BUILD_TYPE} "DEBUG" BUILD_DEBUG)
 option(GSIBUILD "Build only libs for GSI" ON)
 
 set(definitions "UNDERSCORE" "NFILES=32" "MAXCD=250" "MAXNC=600" "MXNAF=3")
```
Is ``definitions`` variable used?

DusanJovic-NOAA commented on CMakeLists.txt:
```diff
@@ -99,30 +105,81 @@ foreach(kind ${kinds})
+#   EXPORT ${PROJECT_NAME}-config
+#   RUNTIME DESTINATION ${CMAKE_INSTALL_PREFIX}/bin
+#   LIBRARY DESTINATION ${CMAKE_INSTALL_PREFIX}/lib
+#   ARCHIVE DESTINATION ${CMAKE_INSTALL_PREFIX}/lib)
```
Can be removed.

DusanJovic-NOAA commented on cmake/GNU.cmake:
```diff
@@ -0,0 +1,68 @@
+
+####################################################
+
+add_definitions(-D__GFORTRAN__)
```
Not necessary. GNU compiler defines this macro. 
https://gcc.gnu.org/onlinedocs/cpp/Common-Predefined-Macros.html

DusanJovic-NOAA commented on CMakeLists.txt:
```diff
@@ -1,22 +1,48 @@
 endif()
 
+set(fortran_8_flags ${FREAL8} ${FINT8})
+set(shared_flags  ${DEBINFO} ${FOPT3} ${TRACEBACK} ${AVX2})
```
WIth GNU compiler I see these warnings;
```
Scanning dependencies of target bufr_8_c
[ 33%] Building C object CMakeFiles/bufr_8_c.dir/src/arallocc.c.o
cc1: warning: command line option ‘-fbacktrace’ is valid for Fortran but not for C
[ 33%] Building C object CMakeFiles/bufr_8_c.dir/src/ardllocc.c.o
cc1: warning: command line option ‘-fbacktrace’ is valid for Fortran but not for C
[ 33%] Building C object CMakeFiles/bufr_8_c.dir/src/bort_exit.c.o
cc1: warning: command line option ‘-fbacktrace’ is valid for Fortran but not for C
```
``${TRACEBACK}`` should be used in ``shared_flags``

DusanJovic-NOAA commented on cmake/GNU.cmake:
```diff
@@ -0,0 +1,68 @@
+set (FOPT2 "-O2")
+set (FOPT3 "-O3")
+set (FOPT4 "-O3")
+set (AVX2 "-O3")
```
``AVX2`` variable should enable the use of AVX2 instructions (``-maxv2``), and should not set optimization level.

DusanJovic-NOAA commented:
If ``double``, ``doubleDA``, and ``supersize`` directories are not used anymore remove them.

mark-a-potts commented:
Just pushed an update that addresses most of these issues. I am not sure what to do with all the different "kinds". The static allocation does not seem to work with GNU, so that build defaults to dynamic for everything. Is there really a need for both Dynamic and Static allocation builds? I believe both 4 byte and double are used, but not sure about any of the others.

DusanJovic-NOAA commented:
I asked several times NCEPLIBS group do we really need ``_8`` libraries. Do we know for sure that there are applications that are built with all integers being promoted to 8 bytes. Why would anyone need **all** integers to be 64 bits wide. I think we should revisit that and hopefully stop building libraries nobody needs.

aerorahul commented on CMakeLists.txt:
```diff
@@ -1,95 +1,75 @@
+set(c_d_flags "-DUNDERSCORE")
+set(fortran_4_flags)
 
 file(GLOB fortran_src ${CMAKE_CURRENT_SOURCE_DIR}/src/*.f ${CMAKE_CURRENT_SOURCE_DIR}/src/*.F)
```
Please do not glob. and move the building of the library into src.
The top level CMakeLists.txt should not do anything more than basic setup and add_subdirectory.

aerorahul commented on CMakeLists.txt:
```diff
@@ -99,30 +79,63 @@ foreach(kind ${kinds})
-    COMPILE_DEFINITIONS "${c_definitions};${c_${kind}_definitions}")
-  
-  set_source_files_properties(${fortran_src} PROPERTIES
+  add_library(${lib_name}_f OBJECT ${fortran_src})
```
Need to explicitly say STATIC.
Also add an ALIAS

aerorahul commented on CMakeLists.txt:
```diff
@@ -99,30 +79,63 @@ foreach(kind ${kinds})
+    COMPILE_DEFINITIONS "${allocation_def}")
+
+  add_library(${lib_name}_c OBJECT ${c_src})
+  set_target_properties(${lib_name}_c PROPERTIES POSITION_INDEPENDENT_CODE ON)
```
PIC is not necessary when library is defined STATIC mode

aerorahul commented on cmake/GNU.cmake:
```diff
@@ -0,0 +1,64 @@
+set (FOPT0 "-O0")
```
This file is not needed.
The last 2 lines are all that are being used.
Please follow one of the other NCEPLIBS- libraries. e.g. nemsio or bacio

aerorahul commented on cmake/Intel.cmake:
```diff
@@ -0,0 +1,62 @@
+set (FOPT0 "-O0")
```
same comment as GNU.cmake

aerorahul commented on cmake/PackageConfig.cmake.in:
```diff
@@ -0,0 +1,14 @@
+
+# Include targets file.  This will create IMPORTED target @PROJECT_NAME@
+include("${CMAKE_CURRENT_LIST_DIR}/@PROJECT_NAME@-targets.cmake")
+
```
Please include the config-version file that CMake produces.

aerorahul commented on cmake/PackageConfig.cmake.in:
```diff
@@ -0,0 +1,14 @@
+check_required_components("@PROJECT_NAME@")
+
+get_target_property(location @PROJECT_NAME@::@PROJECT_NAME@_4 LOCATION)
+message(STATUS "Found @PROJECT_NAME@: ${location} (found version \"@PROJECT_VERSION@\")")
```
`@PROJECT_VERSION@` should be replaced with `${PACKAGE_VERSION}`

mark-a-potts replied:
This now follows the same standard as w3emc which uses PROJECT_VERSION.

aerorahul replied:
yeah. That will change. You can ignore this comment.

aerorahul commented on cmake/osx_extras.cmake:
```diff
@@ -0,0 +1,27 @@
+# OSX fixes/workarounds
```
This file is not used.  Remove.

aerorahul commented on CMakeLists.txt:
```diff
@@ -1,95 +1,75 @@
+add_compile_definitions(definitions)
+
+if(CMAKE_C_COMPILER_ID MATCHES "^(Intel)$")
+  include(${CMAKE_CURRENT_SOURCE_DIR}/cmake/Intel.cmake)
   set(IntelComp true )
```
please remove including these files. Explicitly set flags.  See other repos for how this is done to be consistent with other NCEPLIBS.

aerorahul commented on CMakeLists.txt:
```diff
@@ -1,95 +1,75 @@
+  set(allocation_def "DYNAMIC_ALLOCATION")
 endif()
 
+set(fortran_8_flags ${FREAL8} ${FINT8} ${TRACEBACK})
+set(shared_flags  ${DEBINFO} ${FOPT3} ${AVX2})
+set(fortran_8_DA_flags ${FREAL8} ${FINT8} ${TRACEBACK})
+set(fortran_d_DA_flags ${FREAL8} ${TRACEBACK})
+set(c_SS_flags ${MMMED} ${SHARED})
```
please spell out these explicitly.  They are not that long.

mark-a-potts replied:
They are now defined explicitly for each compiler above.

aerorahul commented on CMakeLists.txt:
```diff
@@ -1,95 +1,75 @@
 
+set(c_4_DA_flags "-DUNDERSCORE")
+set(c_8_DA_definitions "F77_INTSIZE_8")
+set(c_d_DA_flags "-DUNDERSCORE")
```
-DUNDERSCORE is covered under add_definitions(UNDERSCORE)

aerorahul commented on src/CMakeLists.txt:
```diff
@@ -0,0 +1,111 @@
+cmake_minimum_required(VERSION 3.15)
```
not the right place to check this.

aerorahul commented on CMakeLists.txt:
```diff
@@ -1,128 +1,37 @@
   set(GNUComp true )
-elseif(CMAKE_CXX_COMPILER_ID MATCHES "pgc*")
-  set(PGIComp true )
+  set(allocation_def "DYNAMIC_ALLOCATION")
```
move under `src/CMakeLists.txt`

aerorahul commented on CMakeLists.txt:
```diff
@@ -1,128 +1,37 @@
 endif()
 
-if(CMAKE_CXX_COMPILER_ID STREQUAL "Intel")
+if(CMAKE_C_COMPILER_ID MATCHES "^(Intel)$")
```
Still has `IntelComp` and `GNUComp`.  Please remove.

mark-a-potts replied:
IntelComp is used in the kinds loop.

aerorahul commented on CMakeLists.txt:
```diff
@@ -1,128 +1,37 @@
+  set(allocation_def "DYNAMIC_ALLOCATION")
 endif()
 
 STRING(COMPARE EQUAL ${CMAKE_BUILD_TYPE} "RelWithDebInfo" BUILD_RELEASE)
```
Where are these used?
Please remove.

mark-a-potts replied:
allocation_def is used in the kinds loop with IntelComp, but i cleaned up the number of places allocation_def was set.

aerorahul replied:
Ok. Thanks for cleaning up the `allocation_def`.
What about `BUILD_RELEASE`, `BUILD_PRODUCTION` and `BUILD_DEBUG`?
They are just redefinitions.

aerorahul commented on src/CMakeLists.txt:
```diff
@@ -0,0 +1,111 @@
+set(definitions "UNDERSCORE" "NFILES=32" "MAXCD=250" "MAXNC=600" "MXNAF=3")
+add_compile_definitions(definitions)
+
+if(InteComp)
+  set (FREAL8 "-r8")
+  set (FINT8 "-i8")
+  set (TRACEBACK "-traceback")
+  set (MMMED "-mcmodel=medium")
+elseif(GNUComp)
+  set (FREAL8 "-fdefault-double-8" "-fdefault-real-8")
+  set (FINT8 "-fdefault-integer-8")
+  set (TRACEBACK "-fbacktrace")
+  set (MMMED "")
+endif()
```
Please do an explicit check without IntelComp or GNUComp

aerorahul commented on src/list_of_files.cmake:
```diff
@@ -0,0 +1,330 @@
+set(fortran_src 
+  ${CMAKE_CURRENT_SOURCE_DIR}/adn30.f
```
`${CMAKE_CURRENT_SOURCE_DIR}/` is not required. These are in the same directory as the current directory.

mark-a-potts replied:
Agreed.

aerorahul commented on src/CMakeLists.txt:
```diff
@@ -0,0 +1,111 @@
+  # -DDYNAMIC_ALLOCATION when compiled with DA and -DSTATIC_ALLOCATION
+  # check current kind and if it has 'DA' in it then set compiler def
+
+  if(IntelComp)
```
`IntelComp` should be replaced with `CMAKE_C_COMPILER_ID` ...

aerorahul commented on src/CMakeLists.txt:
```diff
@@ -0,0 +1,111 @@
+set(fortran_4_flags)
+
+include("list_of_files.cmake")
+set(kinds "4" "d" "8" "4_DA" "8_DA" "d_DA")
```
Can someone explain where are these compiled libraries used?
`4_DA`, `8_DA`, `d_DA`

aerorahul commented on src/CMakeLists.txt:
```diff
@@ -0,0 +1,111 @@
+set(c_8_DA_definitions "F77_INTSIZE_8")
+set(fortran_4_DA_flags)
+set(c_8_definitions "F77_INTSIZE_8")
+set(fortran_4_flags)
```
empty variables don't need to exist or defined.

aerorahul commented on src/CMakeLists.txt:
```diff
@@ -0,0 +1,111 @@
+set(c_nonDA_allocation_def "STATIC_ALLOCATION")
+
+set(c_8_DA_definitions "F77_INTSIZE_8")
+set(fortran_4_DA_flags)
```
empty variables don't need to be defined or exist

aerorahul commented on src/CMakeLists.txt:
```diff
@@ -0,0 +1,111 @@
+set(c_DA_allocation_def "DYNAMIC_ALLOCATION")
+set(c_nonDA_allocation_def "STATIC_ALLOCATION")
+
+set(c_8_DA_definitions "F77_INTSIZE_8")
```
the values of  `c_8_DA_definitions` is the same as `c_8_definitions`. i.e. it does not distinguish between `DA` and `nonDA`.  This can be reduced to `c_8_definitions`

aerorahul commented on src/CMakeLists.txt:
```diff
@@ -0,0 +1,111 @@
+endif()
+
+
+set(fortran_8_flags ${FREAL8} ${FINT8} ${TRACEBACK})
```
fortran_8_flags are the same as fortran_8_DA_flags.  They can be combined and reduced to fortran_8_flags

mark-a-potts replied:
The variables are looped over in the kind loop, so need to be duplicated for that reason.

aerorahul replied:
Ok. I understand that.

Also, does GNU build all 6 kinds? In the foreach loop, only Intel compiler is checked and then kind is checked.
I don't understand what is going on with these and do we really need these?  What does the GSI use?

aerorahul commented on src/CMakeLists.txt:
```diff
@@ -0,0 +1,111 @@
+
+
+set(fortran_8_flags ${FREAL8} ${FINT8} ${TRACEBACK})
+set(shared_flags  ${DEBINFO} ${FOPT3} ${AVX2})
```
what is DEBINFO, FOPT3, AVX2? 
And why would one use AVX2 with GNU compiler?

mark-a-potts replied:
There are mtune options for the gnu compiler that match up with AVX2

aerorahul replied:
regardless, they are not defined.
Why cant we reduce this whole down to something like this:
```
if(CMAKE_C_COMPILER_ID MATCHES "^(Intel)$")
  set(c_DA_allocation_def "DYNAMIC_ALLOCATION")
  set(c_SA_allocation_def "STATIC_ALLOCATION")
  set(c_8_definitions "F77_INTSIZE_8")
  set(c_8_DA_definitions "F77_INTSIZE_8")
  set(CMAKE_Fortran_FLAGS "-g -traceback")
  set(CMAKE_Fortran_FLAGS_DEBUG "-O0")
  set(CMAKE_Fortran_FLAGS_RELEASE "-O3")
  set(fortran_8_flags "-r8 -i8")
  set(fortran_d_flags "-r8")
  set(fortran_8_DA_flags "-r8 -i8")
  set(fortran_d_DA_flags "-r8")
elseif(CMAKE_C_COMPILER_ID MATCHES "^(GNU|Clang|AppleClang)$")
  set(CMAKE_C_FLAGS_DEBUG "-ggdb -Wall -O0")
  set(CMAKE_Fortran_FLAGS "-g -fbacktrace -funroll-loops")
  set(CMAKE_Fortran_FLAGS_DEBUG "-ggdb -Wall -O0")
  set(CMAKE_Fortran_FLAGS_RELEASE "-O3")
  set(fortran_8_flags "-fdefault-real-8 -fdefault-integer-8")
  set(fortran_d_flags "-fdefault-real-8")
endif()
```

aerorahul commented on src/CMakeLists.txt:
```diff
@@ -0,0 +1,111 @@
+set(shared_flags  ${DEBINFO} ${FOPT3} ${AVX2})
+set(fortran_8_DA_flags ${FREAL8} ${FINT8} ${TRACEBACK})
+set(fortran_d_DA_flags ${FREAL8} ${TRACEBACK})
+set(c_SS_flags ${MMMED} ${SHARED})
```
What is SHARED?

mark-a-potts replied:
c_SS_flags weren't used anymore. I removed them.

aerorahul commented on src/CMakeLists.txt:
```diff
@@ -0,0 +1,111 @@
+set(shared_flags  ${DEBINFO} ${FOPT3} ${AVX2})
+set(fortran_8_DA_flags ${FREAL8} ${FINT8} ${TRACEBACK})
+set(fortran_d_DA_flags ${FREAL8} ${TRACEBACK})
+set(c_SS_flags ${MMMED} ${SHARED})
```
where is c_SS_flags used?

aerorahul commented:
Here is an explanation from @jbathegit (Jeff Ator) on the differences between the DA and  "non"DA variants of the bufr library:

The only variants in the latest version 11.3.0 of the library are *4 (4-byte reals and 4-byte ints), *8 (8-byte reals and 8-byte ints) and *d (8-byte reals and 4-byte ints), and then there's a DA (dynamic allocation) version of each, for a total of 6 library builds.  I have no idea which application programs are using which builds on the WCOSS or elsewhere; rather, we've just supported all of the different variant builds for 20+ years (going back to the advent of the library itself).  But the same version of the source code can be used to build any of the variants, using switches in the make script and conditional compile directives in the source code.

When an application program is linked to a DA build of the BUFRLIB, you do have to make additional calls to function ISETPRM to tell the library how large you want to set certain parameters and array sizes (that's kind of the whole point of being able to dynamically allocate ;-).  The whole point of the DA builds is that it gives flexibility to the end user if, e.g. they have some ginormous BUFR messages, or messages containing unusually large bitmaps or other structures that might otherwise cause a memory fault when using the SA builds which have pre-set array sizes and other parameters. If users are fine with those settings, then there's no reason to use a DA build because then you end up paying an extra run-time price to do the dynamic memory allocation, and thus you'd be better off just using a corresponding SA build.

Now, having said all that, you can link a DA build to an application program without making any calls to ISETPRM.  As explained in the documentation for this function (see https://emc.ncep.noaa.gov/emc/pages/infrastructure/bufrlib/docs/toc/other/#isetprm), if an application program doesn't make any calls to this function, then the library defaults to using the same parameter and array sizes as for the corresponding SA build.  In other words, you only need to call ISETPRM if you want to change one or more parameters from their default settings; otherwise, there's no need to call this function, and then the only difference would be that you'd just have to pay the additional aforementioned price of having all of your memory allocation done at run-time (in the DA build) vs. at compile-time (in the SA build).
